### ISSUE TITLE
Support for libraries that use universal binaries on macOS

### ISSUE THREAD

**Issue by codeanticode** [benfry/processing4#510]
Some libraries may use universal native libraries that bundle both x86 and aarch64 binaries in the same dylib file, such as the video library. This situation is not currently supported by the current variant/platform system (https://github.com/processing/processing-video/issues/199).

**Comment by benfry** [comment]
Haven't implemented it at the moment because there's still libraries out there called “universal” that are 32-bit and PPC, or 32- and 64-bit Intel, and this might just create more confusion.

Perhaps more important, are there places where this actually helps us? The binaries for the video libraries are already huge, and doubling them in size by shipping universal versions doesn't seem like a good idea.

**Comment by codeanticode** [comment]
The latest gstreamer native libs for Mac are being distributed in universal format (x86_64 + arm64), not as separate packages for each arch (unless we build them ourselves). So this would be in principle needed only to support the video library on apple arm64. I'm not aware of other native libs for Mac that are being distributed as universal binaries for x86 and arm.

**Comment by benfry** [comment]
May be a good idea for us to use lipo and extract them ourselves… I don't love it, but this adds extra complexity in the PDE code and doubles the size of the video library binaries in any exported app on the Mac. It has to work for runtime, export, etc as well as checking to for edge cases where old versions exist, etc. It's messy code that I was happy to get rid of in beta 4 or whenever I did all the work to get this cleaned up and working.

If there are more compelling reasons to support universal in the future we could look into adding it, but for now it seems out of balance with the maintenance cost in the PDE and impact on exported apps.

**Comment by codeanticode** [comment]
I see, didn't know about this tool. I tried it and it seems to work, was able to generate separate sets of native gstreamer libraries. So I agree we don't need to add universal support. 

The issue is when I place the macos-aarch64 folder under library, and try running a video library example with the Apple Silicon release of Processing 4.0b8, I get an UnsatisfiedLinkError error that seems related to JNA:

<img width="1199" alt="image" src="https://user-images.githubusercontent.com/62246/178735066-660a9814-1900-411b-a691-f884b5a85752.png">

I then try running the same example building the latest processing4 source, and looks like Processing is looking for x86_64 natives:

<img width="400" alt="image" src="https://user-images.githubusercontent.com/62246/178735729-d77e7f6f-71fb-41dd-9508-792064608d0d.png">

**Comment by neilcsmith-net** [comment]
> The issue is when I place the macos-aarch64 folder under library, and try running a video library example with the Apple Silicon release of Processing 4.0b8, I get an UnsatisfiedLinkError error that seems related to JNA:

Have you updated JNA?  There was quite an old version in there - one we're about to stop supporting.  Also have you testing any of the GStreamer Java examples work?  Currently doing some work / testing around macOS M1 with gst1-java-core so feel free to follow up with me elsewhere on this.

Still think you should stop shipping with the binaries included! :wink:

**Comment by benfry** [comment]
@codeanticode It'd be a good idea to match the JNA version that's shipping with the PDE these days, to avoid additional conflicts should they somehow collide. It should be a fat version (so it can live in the root with the same layout as you have there).

But more to your question, I'm not following exactly how you have things running there—it actually looks like you're using an x86_64 Java VM and it's (rightly) ignoring the `macos-aarch64` folder, and it's complaining because it can't find a `macos-x86_64` subfolder in the video library. So first make sure that you're not accidentally running an x86_64 Java.

**Comment by codeanticode** [comment]
Updating jna.jar to the version in the PDE solves the problem when running the video library from 4.0b8.

With regards to running the video library from the dev version of Processing 4, I had installed Temurin 17.0.2 before and yes seems it was a x86_64 VM. I installed the latest version from Adoptium website, which showed me the aarch64 package by default, OpenJDK17U-jdk_aarch64_mac_hotspot_17.0.3_7, and with that the aarch64 gstreamer native libs load without error.

Next version of the video library will include the updated jna.jar and aarch64 native libs, so we can close the issue, thanks for the suggestions.
